### ISSUE TITLE
Feature request: Capture device's audio output for recognition.

### ISSUE THREAD

**Issue by RokeJulianLockhart** [aleksey-saenko/MusicRecognizer#54]
#### **Problem**

As [this](https://www.reddit.com/r/androidapps/comments/1cdgah4/comment/l1cjnkg/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button#:~:text=No%20unfortunately.-,audile%20can't%20recognize%20system%20audio,-and%20doesn't%20show) explains, Audile https://github.com/aleksey-saenko/MusicRecognizer/releases/download/v1.4.3/Audile_v1.4.3_github.apk currently doesn't appear to be able to recognise sound which hasn't been routed via the speakers of the device and subsequently recognised by the microphone. However, [Shazam](https://www.apkmirror.com/apk/apple-inc/shazam/shazam-14-21-0-release/shazam-find-music-concerts-14-21-0-240411-android-apk-download/) is able to.

#### **Solution**

Consequently, please implement the ability to recognise audio from the device which doesn't exit a speaker, whether that's due to 0% audio volume or connected headphones.

**Comment by okgamr** [comment]
Audile already does that. i just tried it and it seems to take the music that is playing on the phone (even with headphones). at least @aleksey-saenko should have known what he built
someone end this issue up

**Comment by RokeJulianLockhart** [comment]
> https://github.com/aleksey-saenko/MusicRecognizer/issues/54#issuecomment-2155228912

@okgamr, I don't understand that last sentence:

> at least @aleksey-saenko should have known what he built someone end this issue up

Can you rephrase it?

**Comment by okgamr** [comment]
i mean, he didn't know this was already working. he edited the title and added the label (or it may have been a bot, idk), so he could just snap in and say "already implemented"

**Comment by aleksey-saenko** [comment]
@okgamr @RokeJulianLockhart 
This request refers to the ability to recognize music when it is played **only** through headphones. This feature has not yet been implemented.
Currently, the app recognizes audio recorded using a microphone. It doesn't matter if the music is playing from the phone or another device, as long as it can be heard by the microphone.
To implement this function, in addition to the microphone, it would also be necessary to capture audio directly from the device. The problem is that Android prohibits doing this without explicit one-time permission. If this were implemented using MediaProjection (an API specifically designed for this purpose that guarantees sound quality), then with each recognition request you will need to grant the permission in system dialog, which is not user-friendly. And you can notice that Shazam does not request such permissions. They likely record audio through the Visualizer, but it is not intended for this purpose, and the recording quality turns out to be very poor (at least in my experiments). Therefore, this feature has not yet found its implementation.

**Comment by aleksey-saenko** [comment]
@okgamr 
If you were listening to music **only** through headphones but the app still recognized it, I assume that your microphone was able to catch this quiet sound (open-type headphones or with built-in mic, high volume?), amplify it, and that was enough for the service. You can perform some tests with the internet turned off; in this case, recordings will be saved in queue, and you'll be able to listen to their quality.

**Comment by okgamr** [comment]
> @okgamr If you were listening to music **only** through headphones but the app still recognized it, I assume that your microphone was able to catch this quiet sound (open-type headphones or with built-in mic, high volume?), amplify it, and that was enough for the service. You can perform some tests with the internet turned off; in this case, recordings will be saved in queue, and you'll be able to listen to their quality.

ok.... then i will just use the headphones at low volume i guess

**Comment by okgamr** [comment]
i'm dreaming, i'm day-dreaming. does it recognizes by a built-in microphone? insane, that's insane


so, my volume was about 1%, and i put my hand on the mics. nothing

you won

**Comment by 0-BlackSpectrum-0** [comment]
Any news when this feature is going to be implemented?
I came here to ask for this!
Please add this feature

**Comment by aleksey-saenko** [comment]
Added in version 1.7.0, will be available on F-Droid within a few days.

As I mentioned earlier, Android offers only one official way to record the device's audio output - MediaProjection API.
This method has several limitations:
- The function is only available on Android 10+
- Each request to capture device audio requires confirmation in a system dialog
- Playing apps can restrict capturing their audio playback (capturing is allowed by default); in this case, Audile receives silence
- Audile and playing apps must be in the same user profile

But overall it works.

**Comment by aleksey-saenko** [comment]
Version 1.9.0 is now available on F-Droid. This update adds an experimental device audio source, that avoids all the limitations mentioned above.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/aleksey-saenko/MusicRecognizer/issues/54#issuecomment-2848623330

@aleksey-saenko, what did you do differently?

**Comment by aleksey-saenko** [comment]
@RokeJulianLockhart 
The first added audio source uses [MediaProjection](https://developer.android.com/reference/android/media/projection/MediaProjection) API. This source is still available and remains the default option.
The newly added audio source has a completely different implementation, it's based on [Visualizer](https://developer.android.com/reference/android/media/audiofx/Visualizer) API.
The user can choose between these two options.

As I mentioned earlier, Visualizer API is not intended for recording device audio. But it turned out to be possible using some [tricks](https://github.com/aleksey-saenko/MusicRecognizer/blob/master/core/audio/src/main/java/com/mrsep/musicrecognizer/core/audio/audiorecord/soundsource/VisualizerSoundSource.kt).
I'm not sure if this will work on all devices. And most likely, in some future Android version, Visualizer capabilities will be restricted for privacy reasons (though this would also affect Shazam, as I assume they use a similar method). But so far it works.
